source/validate_decorations.cpp

```diff
@@ -225,6 +518,62 @@ spv_result_t CheckDescriptorSetArrayOfArrays(ValidationState_t& vstate) {
   return SPV_SUCCESS;
 }
 
+spv_result_t CheckDecorationsOfBuffers(ValidationState_t& vstate) {
+  for (const auto& def : vstate.all_definitions()) {
+    const auto inst = def.second;
+    const auto words = inst->words();
+    if (SpvOpVariable == inst->opcode() &&
+        (SpvStorageClassUniform == words[3] ||
+         SpvStorageClassPushConstant == words[3])) {
+      const auto ptrInst = vstate.FindDef(words[1]);
+      assert(SpvOpTypePointer == ptrInst->opcode());
+      const auto id = ptrInst->words()[3];
+      if (SpvOpTypeStruct != vstate.FindDef(id)->opcode()) continue;
+      for (const auto& dec : vstate.id_decorations(id)) {
+        const bool isBlock = SpvDecorationBlock == dec.dec_type();
+        const bool isBufferBlock = SpvDecorationBufferBlock == dec.dec_type();
+        if (isBlock || isBufferBlock) {
+          std::string dec_str = isBlock ? "Block" : "BufferBlock";
+          if (isMissingOffsetInStruct(id, vstate)) {
+            return vstate.diag(SPV_ERROR_INVALID_ID)
+                   << "Structure id " << id << " decorated as " << dec_str
+                   << " must be explicitly laid out with Offset decorations.";
+          } else if (hasDecoration(id, SpvDecorationGLSLShared, vstate)) {
+            return vstate.diag(SPV_ERROR_INVALID_ID)
+                   << "Structure id " << id << " decorated as " << dec_str
+                   << " must not use GLSLShared decoration.";
+          } else if (hasDecoration(id, SpvDecorationGLSLPacked, vstate)) {
+            return vstate.diag(SPV_ERROR_INVALID_ID)
+                   << "Structure id " << id << " decorated as " << dec_str
+                   << " must not use GLSLPacked decoration.";
+          } else if (!checkForRequiredDecoration(id, SpvDecorationArrayStride,
```

Comment:
This is going to miss a case.

When the member is an array of matrices, or an array of an array of matrices, etc., then the MatrixStride is required on the struct member that is the outermost array type.

Example vertex shader in GLSL:

```
#version 450

layout (set=1, binding=2) uniform S { mat4x4 arr[2][3]; } B;

void main() {
  vec4 v = B.arr[1][2][1];
}
```
Here's the SPIR-V:
```
; SPIR-V
; Version: 1.0
; Generator: Google Shaderc over Glslang; 7
; Bound: 26
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main"
               OpSource GLSL 450
               OpSourceExtension "GL_GOOGLE_cpp_style_line_directive"
               OpSourceExtension "GL_GOOGLE_include_directive"
               OpName %main "main"
               OpName %v "v"
               OpName %S "S"
               OpMemberName %S 0 "arr"
               OpName %B "B"
               OpDecorate %_arr_mat4v4float_uint_3 ArrayStride 64
               OpDecorate %_arr__arr_mat4v4float_uint_3_uint_2 ArrayStride 192
               OpMemberDecorate %S 0 ColMajor
               OpMemberDecorate %S 0 Offset 0
               OpMemberDecorate %S 0 MatrixStride 16 ;;;;; This is the surprising part
               OpDecorate %S Block
               OpDecorate %B DescriptorSet 1
               OpDecorate %B Binding 2
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
      %float = OpTypeFloat 32
    %v4float = OpTypeVector %float 4
%_ptr_Function_v4float = OpTypePointer Function %v4float
%mat4v4float = OpTypeMatrix %v4float 4
       %uint = OpTypeInt 32 0
     %uint_3 = OpConstant %uint 3
%_arr_mat4v4float_uint_3 = OpTypeArray %mat4v4float %uint_3
     %uint_2 = OpConstant %uint 2
%_arr__arr_mat4v4float_uint_3_uint_2 = OpTypeArray %_arr_mat4v4float_uint_3 %uint_2
          %S = OpTypeStruct %_arr__arr_mat4v4float_uint_3_uint_2
%_ptr_Uniform_S = OpTypePointer Uniform %S
          %B = OpVariable %_ptr_Uniform_S Uniform
        %int = OpTypeInt 32 1
      %int_0 = OpConstant %int 0
      %int_1 = OpConstant %int 1
      %int_2 = OpConstant %int 2
%_ptr_Uniform_v4float = OpTypePointer Uniform %v4float
       %main = OpFunction %void None %3
          %5 = OpLabel
          %v = OpVariable %_ptr_Function_v4float Function
         %24 = OpAccessChain %_ptr_Uniform_v4float %B %int_0 %int_1 %int_2 %int_1
         %25 = OpLoad %v4float %24
               OpStore %v %25
               OpReturn
               OpFunctionEnd
```
Note where the MatrixStride is applied: It's on member 0 of the struct, not on the matrix type.

I have filed #1637 to track this issue.